[JSC] Loop unrolling

hyjorc1 opened the pull request:
#### 22ac06375c5a827734a281c528a643815c847589
<pre>
[JSC] Loop unrolling
<a href="https://bugs.webkit.org/show_bug.cgi?id=265884">https://bugs.webkit.org/show_bug.cgi?id=265884</a>
<a href="https://rdar.apple.com/119195561">rdar://119195561</a>

Reviewed by Yusuke Suzuki and Justin Michaud.

This patch, based on Justin&apos;s previous draft patch[1], adds a new loop
unrolling optimization phase to the FTL tier. The loop unrolling phase
enhances performance by reducing loop overhead in tightly nested loops
with a limited number of iterations, while maintaining the correctness
of the execution graph.

Changes:
* Introduced a new DFGLoopUnrollingPhase to optimize small, simple loops
  by unrolling them during DFG JIT compilation.
* Added infrastructure improvements in Graph, BasicBlock, and Node to
  facilitate cloning and unrolling.
* Integrated the phase into the DFG pipeline with configurable options
  (useLoopUnrolling, verboseLoopUnrolling, printEachUnrolledLoop).
* Added benchmarks to validate correctness and performance gains.

Future Work:
* Extend support to complex loops with nested structures and advanced
  control flows. See micro-benchmarks tests in [1].
* Enhance unrolling profitability metrics and speculative optimizations
  for dynamic iteration counts.
* Measure performance impact in real-world scenarios and expand profiling
  capabilities.

Microbenchmark Results:
                                              before                    after

Benchmarks with progressions:

loop-unrolling-1                          2.7693+-0.0279     ^      2.5839+-0.0396        ^ definitely 1.0718x faster
loop-unrolling-2                          3.4357+-0.0358     ^      2.7861+-0.0379        ^ definitely 1.2332x faster
loop-unrolling-3                          2.9092+-0.0304     ^      2.6302+-0.0362        ^ definitely 1.1061x faster

Future work benchmarks:

loop-unrolling-4                          3.1286+-0.0442            3.0581+-0.0490          might be 1.0230x faster
loop-unrolling-array-clone-small          0.8926+-0.0082            0.8918+-0.0082
loop-unrolling-core-sha1                  0.3992+-0.0066            0.3919+-0.0071          might be 1.0188x faster
loop-unrolling-variable-medium            5.9324+-0.0327            5.8792+-0.0220
loop-unrolling-variable-medium-dep        8.3997+-0.7200            8.1726+-0.6305          might be 1.0278x faster
loop-unrolling-array-clone-big            0.8996+-0.0095            0.8941+-0.0099
loop-unrolling-constant-small             0.4556+-0.0080            0.4506+-0.0119          might be 1.0111x faster
loop-unrolling-variable-large            53.8354+-0.0943           53.7705+-0.0802
loop-unrolling-variable-small             0.4518+-0.0075            0.4502+-0.0078

[1] <a href="https://github.com/WebKit/WebKit/pull/21342">https://github.com/WebKit/WebKit/pull/21342</a>

* JSTests/microbenchmarks/loop-unrolling-1.js: Added.
(assert):
(test):
* JSTests/microbenchmarks/loop-unrolling-2.js: Added.
(assert):
(test):
* JSTests/microbenchmarks/loop-unrolling-3.js: Added.
(assert):
(test):
* JSTests/microbenchmarks/loop-unrolling-4.js: Added.
(assert):
(test):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3SparseCollection.h:
(JSC::B3::SparseCollection::addNew):
* Source/JavaScriptCore/dfg/DFGBasicBlock.h:
* Source/JavaScriptCore/dfg/DFGBasicBlockInlines.h:
(JSC::DFG::BasicBlock::appendNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp: Added.
(JSC::DFG::LoopUnrollingPhase::LoopData::header const):
(JSC::DFG::LoopUnrollingPhase::LoopData::condition const):
(JSC::DFG::LoopUnrollingPhase::LoopData::isInductionVariable):
(JSC::DFG::LoopUnrollingPhase::LoopUnrollingPhase):
(JSC::DFG::LoopUnrollingPhase::run):
(JSC::DFG::LoopUnrollingPhase::selectDeepestNestedLoop):
(JSC::DFG::LoopUnrollingPhase::tryUnroll):
(JSC::DFG::LoopUnrollingPhase::locatePreHeader):
(JSC::DFG::LoopUnrollingPhase::locateTail):
(JSC::DFG::LoopUnrollingPhase::identifyInductionVariable):
(JSC::DFG::LoopUnrollingPhase::isValid):
(JSC::DFG::LoopUnrollingPhase::makeBlock):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
(JSC::DFG::performLoopUnrolling):
(JSC::DFG::LoopUnrollingPhase::LoopData::dump const):
(JSC::DFG::LoopUnrollingPhase::isNodeCloneable):
(JSC::DFG::LoopUnrollingPhase::cloneNode):
(JSC::DFG::LoopUnrollingPhase::cloneNodeImpl):
(JSC::DFG::LoopUnrollingPhase::isSupportedConditionOp):
(JSC::DFG::LoopUnrollingPhase::isSupportedUpdateOp):
(JSC::DFG::std::function&lt;bool):
(JSC::DFG::std::function&lt;CheckedInt32):
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.h: Copied from Source/JavaScriptCore/dfg/DFGBasicBlockInlines.h.
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::setOpInfo):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/287398@main">https://commits.webkit.org/287398@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4429054f93d65ea9e991d385dd3e6a3facb3bf6c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79378 "21 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58405 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83987 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30513 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67470 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6667 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62087 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19970 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82445 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52144 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72321 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42396 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49488 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26508 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28917 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72543 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70674 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26955 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85392 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78566 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6653 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4635 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70339 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68198 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69586 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17367 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13633 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12498 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100909 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6605 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24621 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6518 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9993 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8316 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,822 @@
+
+        // Mapping from the origin to the clones.
+        HashMap<BasicBlock*, BasicBlock*> blockClones;
+        HashMap<Node*, Node*> nodeClones;
```
Use `UncheckedKeyHashMap`

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,822 @@
+
+class LoopUnrollingPhase : public Phase {
+public:
+    bool verbose = Options::verboseLoopUnrolling();
```
I like keeping this as constant and, when using `Options::verboseLoopUnrolling()`, using it directly, like,
```
dataLogLnIf(Options::verboseLoopUnrolling(), "...");
```

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,822 @@
+    }
+}
+
+std::function<bool(CheckedInt32, CheckedInt32)> LoopUnrollingPhase::comparisonFunction(Node* condition)
```
Let's just return a function pointer as you can cast non-capturing lambda to a function pointer, like,
```
bool(*)(CheckedInt32, CheckedInt32)
```

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,822 @@
+    bool isSupportedConditionOp(NodeType op);
+    bool isSupportedUpdateOp(NodeType op);
+    std::function<bool(CheckedInt32, CheckedInt32)> comparisonFunction(Node* condition);
+    std::function<CheckedInt32(CheckedInt32, CheckedInt32)> updateFunction(Node* update);
```
Let's avoid using `std::function`. This allocates object in system malloc and not efficient. In this case, we should just use function pointer.

Constellation commented on Source/JavaScriptCore/dfg/DFGBasicBlock.h:
```diff
@@ -163,6 +163,8 @@ struct BasicBlock : RefCounted<BasicBlock> {
     void removePredecessor(BasicBlock* block);
     void replacePredecessor(BasicBlock* from, BasicBlock* to);
 
+    Node* appendNode(Graph&, Node&&);
```
Since moving DFG node is not possible (adding / removing take special cares), I suggest making it,
```
Node* appendClonedNode(Graph&, Node*);
```

Constellation commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -217,7 +229,13 @@ class Graph final : public virtual Scannable {
     }
-    
+
+    Node* addNode(Node&& node)
```
Take `const Node& clonedNode`

Constellation commented on Source/JavaScriptCore/b3/B3SparseCollection.h:
```diff
@@ -62,6 +62,11 @@ class SparseCollection {
+    T* addNew(T&& value)
+    {
+        return add(makeUnique<T>(WTFMove(value)));
+    }
```
Let's change it to

```
T* addNew(const T& value)
{
    return add(makeUnique<T>(value));
}
```

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+
+    using NaturalLoop = CPSNaturalLoop;
+
+    typedef bool (*ComparisonFunction)(CheckedInt32, CheckedInt32);
```
You can use `using ComparisonFunction = bool(*)(CheckedInt32, CheckedInt32);` for newer code.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+    using NaturalLoop = CPSNaturalLoop;
+
+    typedef bool (*ComparisonFunction)(CheckedInt32, CheckedInt32);
+    typedef CheckedInt32 (*UpdateFunction)(CheckedInt32, CheckedInt32);
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+    out.print("iterationCount=", iterationCount);
+}
+
+bool LoopUnrollingPhase::isNodeCloneable(HashSet<Node*>& cloneAble, Node* node)
```
Name `clonAble` as `cloneableCache`.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+        Node* clone = cloneAndAppend(node);
+        clone->children.setFirstChild(firstChild);
+        return clone;
+    }
```
I wonder if we can generalize this for `VarArgs` node case. For example, `CheckVarargs` sounds quite similar.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+public:
+    static constexpr uint32_t MaxUnrolledLoopCount = 2;
+    static constexpr uint32_t MaxIterationCount = 4;
+    static constexpr uint32_t MaxLoopBodySize = 6;
```
For constant variable, we start with lower case. like, `maxUnrolledLoopCount`.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+        return true;
+    }
+
+    bool isValid(const NaturalLoop* loop)
```
Let's rename this `isValid` to something more descriptive. Currently, it is unknown what "valid" is.
What does this function check? Let's name this function with that condition.

hyjorc1 replied:
Split the `isValid` to `shouldUnrollLoop` and `canCloneLoop`, where `shouldUnrollLoop` is checking block reachability and total loop node count and `canCloneLoop` is checking the loop node cloneability.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+        for (unsigned loopIndex = m_graph.m_cpsNaturalLoops->numLoops(); loopIndex--;) {
+            const NaturalLoop& loop = m_graph.m_cpsNaturalLoops->loop(loopIndex);
+
+            if (loop.size() > MaxLoopBodySize) {
```
Maybe we should also check # of DFG nodes in this loop? Currently, it is checking # of basic blocks in this loop, and it is not directly corresponding to the size of body (# of DFG nodes).

hyjorc1 replied:
Replaced `MaxLoopBodySize` with `maxLoopNodeSize` and move the check into `shouldUnrollLoop`.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+        const NaturalLoop* target = nullptr;
+        int32_t maxDepth = INT_MIN;
+
+        UncheckedKeyHashMap<const NaturalLoop*, int32_t> depthCache;
```
NaturalLoop has sequential `index()`. Thus probably we can just use `IndexMap`.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,818 @@
+            dataLogLnIf(Options::verboseLoopUnrolling(), "\nTry unroll innerMostLoop=", *loop, " with innerMostOuterLoop=", outerLoop ? *outerLoop : NaturalLoop());
+        }
+
+        if (!isValid(loop))
```
Let's rename it to more descriptive function.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,823 @@
+            int32_t depth = 0;
+            const NaturalLoop* current = &loop;
+            while (current) {
+                int32_t& cachedDepth = depthCache[current->index()];
```
It does not need to be reference. Use `int32_t` instead of `int32_t&`.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,823 @@
+    case CompareLessEq:
+    case CompareGreater:
+    case CompareGreaterEq:
+    case CompareEq:
```
How about adding `CompareStrictEq`?

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,823 @@
+    case CompareGreaterEq:
+        return [](auto a, auto b) { return a >= b; };
+    case CompareEq:
+        return [](auto a, auto b) { return a == b; };
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,823 @@
+                        continue;
+                    dataLogLnIf(Options::verboseLoopUnrolling(), "Induction variable ", data.inductionVariable->index(), " is updated at node ", node->index(), " at ", *body);
+                    ++updateCount;
+                    if (updateCount != 1)
+                        return false;
```
Probably worth extending in the future.

hyjorc1 replied:
Yeah, that could be tricky for example 
```
for (let i = 0; i < 4; ) {
    ...
    if (...) {
         i += 1;
    } else {
         i += 2;
    }
}
```
And that makes it difficult to extract a constant number of iterations for the loop. The reason why this simple loop unrolling with constant iterations works is that
1. Only one update to the induction variable.
2. That update must be traceable/reachable from the terminal (branch) in the tail block.

And this one doesn't work as well since the terminal (`i < 4`) gets the induction variable `i` from a `Phi` but not from a directly reachable update e.g. `i++`.
```
for (let i = 0; i < 4; ) {
    ...
    if (...) {
         i += 1;
    }
}
```

hyjorc1 replied:
Both examples shown above make loop unrolling with constant iterations hard to predict.

hyjorc1 replied:
Added a FIXME there.

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,823 @@
+    {
+        Node* condition = data.condition();
+        auto isConditionValid = [&]() ALWAYS_INLINE_LAMBDA {
+            if (!isSupportedConditionOp(condition->op()))
```
Should we check whether the condition is typed with `Int32Use`?

hyjorc1 replied:
```
for (i = initialValue; condition(i, operand); i = update(i, updateValue)) { ... }
```
Sorry, I know it's confusing. Actually, `condition` is the first child of the terminal (`Branch`) in the tail block. So, the use kind should be `BooleanUse`. I think you want to say that we want to ensure the condition left/child1 (`i`) and right/child2 (`operand`) are in `Int32Use`.

Since we also want to avoid the unpredictable (actually hard to predict) constant iterations in this example
```
for (let i = 0; i < 4; ) {
    ...
    if (...) {
         i += 1;
    }
}
```
, we should ensure the pattern that condition left (`i`) is exactly the update `update(i, updateValue)`.

hyjorc1 replied:
In this case, `D@35` is the `condition` node, `D@31` is the `update` node, and `D@34` is the `operand` node.
```
 10  1 19:   D@31:< 1:->	ArithAdd(Int32:D@28, Int32:D@29, Int32|PureNum|NeedsNaNOrInfinity|UseAsOther|ReallyWantsInt, Int32, CheckOverflow, Exits, bc#16, ExitValid)
 11  1 19:   D@32:<!0:->	MovHint(Check:Untyped:D@31, MustGen, loc5, W:SideState, ClobbersExit, bc#16, ExitInvalid)
 12  1 19:   D@33:< 1:->	SetLocal(Int32:D@31, loc5(I<Int32>/FlushedInt32), W:Stack(loc5), bc#16, exit: bc#19, ExitValid)  predicting Int32
 13  1 19:   D@34:< 1:->	JSConstant(JS|PureNum|NeedsNaNOrInfinity|UseAsOther, NonBoolInt32, Int32: 4, bc#19, ExitValid)
 14  1 19:   D@35:< 1:->	CompareLess(Int32:D@31, Int32:D@34, Boolean|PureNum|NeedsNegZero|NeedsNaNOrInfinity|UseAsOther, Bool, bc#19, ExitValid)
 15  1 19:   D@36:<!0:->	Branch(KnownBoolean:D@35, MustGen, T:#1/w:10.000000, F:#2/w:1.000000, W:SideState, bc#19, ExitInvalid)
```

Constellation commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -0,0 +1,823 @@
+
+            // Condition left
+            Node* update = condition->child1().node();
+            if (!isSupportedUpdateOp(update->op()))
```
Ditto.

Constellation reviewed (approved):
r=me with comments. Great work

hyjorc1 committed:
[JSC] Loop unrolling
https://bugs.webkit.org/show_bug.cgi?id=265884
rdar://119195561

Reviewed by Yusuke Suzuki and Justin Michaud.

This patch, based on Justin's previous draft patch[1], adds a new loop
unrolling optimization phase to the FTL tier. The loop unrolling phase
enhances performance by reducing loop overhead in tightly nested loops
with a limited number of iterations, while maintaining the correctness
of the execution graph.

Changes:
* Introduced a new DFGLoopUnrollingPhase to optimize small, simple loops
  by unrolling them during DFG JIT compilation.
* Added infrastructure improvements in Graph, BasicBlock, and Node to
  facilitate cloning and unrolling.
* Integrated the phase into the DFG pipeline with configurable options
  (useLoopUnrolling, verboseLoopUnrolling, printEachUnrolledLoop).
* Added benchmarks to validate correctness and performance gains.

Future Work:
* Extend support to complex loops with nested structures and advanced
  control flows. See micro-benchmarks tests in [1].
* Enhance unrolling profitability metrics and speculative optimizations
  for dynamic iteration counts.
* Measure performance impact in real-world scenarios and expand profiling
  capabilities.

Microbenchmark Results:
                                              before                    after

Benchmarks with progressions:

loop-unrolling-1                          2.7693+-0.0279     ^      2.5839+-0.0396        ^ definitely 1.0718x faster
loop-unrolling-2                          3.4357+-0.0358     ^      2.7861+-0.0379        ^ definitely 1.2332x faster
loop-unrolling-3                          2.9092+-0.0304     ^      2.6302+-0.0362        ^ definitely 1.1061x faster

Future work benchmarks:

loop-unrolling-4                          3.1286+-0.0442            3.0581+-0.0490          might be 1.0230x faster
loop-unrolling-array-clone-small          0.8926+-0.0082            0.8918+-0.0082
loop-unrolling-core-sha1                  0.3992+-0.0066            0.3919+-0.0071          might be 1.0188x faster
loop-unrolling-variable-medium            5.9324+-0.0327            5.8792+-0.0220
loop-unrolling-variable-medium-dep        8.3997+-0.7200            8.1726+-0.6305          might be 1.0278x faster
loop-unrolling-array-clone-big            0.8996+-0.0095            0.8941+-0.0099
loop-unrolling-constant-small             0.4556+-0.0080            0.4506+-0.0119          might be 1.0111x faster
loop-unrolling-variable-large            53.8354+-0.0943           53.7705+-0.0802
loop-unrolling-variable-small             0.4518+-0.0075            0.4502+-0.0078

[1] https://github.com/WebKit/WebKit/pull/21342

* JSTests/microbenchmarks/loop-unrolling-1.js: Added.
(assert):
(test):
* JSTests/microbenchmarks/loop-unrolling-2.js: Added.
(assert):
(test):
* JSTests/microbenchmarks/loop-unrolling-3.js: Added.
(assert):
(test):
* JSTests/microbenchmarks/loop-unrolling-4.js: Added.
(assert):
(test):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3SparseCollection.h:
(JSC::B3::SparseCollection::addNew):
* Source/JavaScriptCore/dfg/DFGBasicBlock.h:
* Source/JavaScriptCore/dfg/DFGBasicBlockInlines.h:
(JSC::DFG::BasicBlock::appendNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp: Added.
(JSC::DFG::LoopUnrollingPhase::LoopData::header const):
(JSC::DFG::LoopUnrollingPhase::LoopData::condition const):
(JSC::DFG::LoopUnrollingPhase::LoopData::isInductionVariable):
(JSC::DFG::LoopUnrollingPhase::LoopUnrollingPhase):
(JSC::DFG::LoopUnrollingPhase::run):
(JSC::DFG::LoopUnrollingPhase::selectDeepestNestedLoop):
(JSC::DFG::LoopUnrollingPhase::tryUnroll):
(JSC::DFG::LoopUnrollingPhase::locatePreHeader):
(JSC::DFG::LoopUnrollingPhase::locateTail):
(JSC::DFG::LoopUnrollingPhase::identifyInductionVariable):
(JSC::DFG::LoopUnrollingPhase::isValid):
(JSC::DFG::LoopUnrollingPhase::makeBlock):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
(JSC::DFG::performLoopUnrolling):
(JSC::DFG::LoopUnrollingPhase::LoopData::dump const):
(JSC::DFG::LoopUnrollingPhase::isNodeCloneable):
(JSC::DFG::LoopUnrollingPhase::cloneNode):
(JSC::DFG::LoopUnrollingPhase::cloneNodeImpl):
(JSC::DFG::LoopUnrollingPhase::isSupportedConditionOp):
(JSC::DFG::LoopUnrollingPhase::isSupportedUpdateOp):
(JSC::DFG::std::function<bool):
(JSC::DFG::std::function<CheckedInt32):
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.h: Copied from Source/JavaScriptCore/dfg/DFGBasicBlockInlines.h.
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::setOpInfo):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/287398@main
